### ISSUE TITLE
Camera extrinsics differ significantly after reset despite identical robot and camera configuration

### ISSUE THREAD

**Issue by ZQuang2202** [haosulab/ManiSkill#1164]
Hi ManiSkill team,

I am encountering an issue where the camera extrinsics differ by a large amount after each environment reset using different seeds, even though I explicitly set the robot arm joints and the camera pose identically every time.

According to the original code, the seed is used to initialize different objects in the tabletop environment.

Below is the code I used to reproduce the issue:

```python
env_id = "PickSingleYCB-v1"   # Pick Single YCB tabletop environment
env = gym.make(
    env_id,
    obs_mode='rgbd',
    control_mode="pd_joint_pos",
    render_mode="rgb_array",
    reward_mode=None,
    sensor_configs=dict(shader_pack='default'),
    human_render_camera_configs=dict(shader_pack='default'),
    viewer_camera_configs=dict(shader_pack='default'),
    sim_backend='gpu',
    enable_shadow=True
)

env = RecordEpisode(
    env,
    output_dir='pick_YCB_Single',
    save_trajectory=False,
    trajectory_name='abc',
    save_video=True,
    source_type="motionplanning",
    source_desc="motionplanning solution PickSingleYCB-V1",
    video_fps=20,
    save_on_reset=False
)

seed = 32  # Different seed for different object
obs, _ = env.reset(seed=seed)

camera_params = obs["sensor_param"]['hand_camera']
extrinsic_cv = camera_params['extrinsic_cv'][0].cpu().numpy()
print("Extrinsic CV:\n", extrinsic_cv)
intrinsic_cv = camera_params['intrinsic_cv'][0].cpu().numpy()
print("Intrinsic CV:\n", intrinsic_cv)
cam2world = camera_params['cam2world_gl'][0].cpu().numpy()
print("Cam2World:\n", cam2world)
```

**Output:**

```
##### seed = 32
Extrinsic CV:
 [[ 0.01867748 -0.99979615 -0.00766668  0.03890251]
 [-0.999095   -0.01837022 -0.03836173 -0.00177538]
 [ 0.03821307  0.00837624 -0.9992345   0.41091484]]

Intrinsic CV:
 [[192.   0. 192.]
 [  0. 192. 192.]
 [  0.   0.   1.]]

Cam2World:
 [[ 0.01867735  0.9990952  -0.0382131  -0.01820273]
 [-0.9997963   0.01837009 -0.00837624  0.03542005]
 [-0.00766668  0.03836175  0.9992345   0.4108304 ]
 [ 0.          0.          0.          1.        ]]


##### seed = 5
Extrinsic CV:
 [[-0.03497105 -0.999373   -0.00553296  0.00673488]
 [-0.9990251   0.03480856  0.02714944 -0.03462344]
 [-0.02693987  0.00647702 -0.999616    0.37865257]]
Intrinsic CV:
 [[192.   0. 192.]
 [  0. 192. 192.]
 [  0.   0.   1.]]
Cam2World:
 [[-0.034971    0.9990251   0.02693984 -0.02415334]
 [-0.99937296 -0.03480852 -0.00647697  0.00548331]
 [-0.00553292 -0.02714946  0.9996161   0.37948456]
 [ 0.          0.          0.          1.        ]]
```

---

Furthermore, I noticed discrepancies when I computed `cam2world_gl` manually from the provided `extrinsic_cv`.

For example, with seed = 32, I used the following conversion:

```python
R = extrinsic_cv[:3, :3]
t = extrinsic_cv[:3, 3]

T_cam2world = np.eye(4)
T_cam2world[:3, :3] = R.T
T_cam2world[:3, 3] = -R.T @ t
```
This produced:

```python
array([[ 0.01867748, -0.99909502,  0.03821307, -0.0182027 ],
       [-0.99979615, -0.01837022,  0.00837624,  0.03542004],
       [-0.00766668, -0.03836173, -0.9992345 ,  0.41083044],
       [ 0.        ,  0.        ,  0.        ,  1.        ]])
```

There is a noticeable difference in the sign and orientation compared to the `cam2world_gl` returned by the environment.

---

**Questions:**

* Is this discrepancy expected?
* Why does the camera extrinsic matrix differ significantly between resets when only the seed for the object initialization changes (with robot pose and camera pose remaining constant)?
* How can I ensure that the camera extrinsics remain consistent and deterministic across resets?

Any guidance on how to correctly compute or fix the camera extrinsic parameters would be greatly appreciated.

Thank you very much for your help!

---

**Environment Details:**

* Environment: `PickSingleYCB-v1`
* ManiSkill version: *(3.0.0b14)*
* Python version: *(e.g., 3.9)*
* OS: *(e.g., Ubuntu 20.04)*
* Simulation backend: GPU with shadows enabled

**Comment by StoneT2000** [comment]
By default most table top environments also randomize robot joint positions. If the robot joints change, the hand camera position also change hence the change in extrinsics.

You might notice in the environment code there is a call to initialize the table scene, which includes this randomization

**Comment by ZQuang2202** [comment]
> By default most table top environments also randomize robot joint positions. If the robot joints change, the hand camera position also change hence the change in extrinsics.
> 
> You might notice in the environment code there is a call to initialize the table scene, which includes this randomization

```python
def _initialize_episode(self, env_idx: torch.Tensor, options: dict):
        with torch.device(self.device):
            b = len(env_idx)
            self.scene_builder.initialize(env_idx)
            goal_pos = torch.rand(size=(b, 3)) * torch.tensor(
                [0.3, 0.5, 0.1]
            ) + torch.tensor([-0.15, -0.25, 0.35])
            self.goal_pos = goal_pos
            self.goal_site.set_pose(Pose.create_from_pq(self.goal_pos))

            # reset objects to original poses
            if b == self.num_envs:
                # if all envs reset
                self.all_objects.pose = self.all_objects.initial_pose
            else:
                # if only some envs reset, we unfortunately still have to do some mask wrangling
                mask = torch.isin(self.all_objects._scene_idxs, env_idx)
                self.all_objects.pose = self.all_objects.initial_pose[mask]
            
            # Initialize robot arm to a higher position above the table than the default typically used for other table top tasks
            if self.robot_uids == "panda" or self.robot_uids == "panda_wristcam":
                # fmt: off
                qpos = np.array(
                    [0.0, 0, 0, -np.pi * 2 / 3, 0, np.pi * 2 / 3, np.pi / 4, 0.04, 0.04]
                )
                # fmt: on
                qpos[:-2] += self._episode_rng.normal(
                    0, self.robot_init_qpos_noise, len(qpos) - 2
                )
                self.agent.reset(qpos)
                self.agent.robot.set_root_pose(sapien.Pose([-0.615, 0, 0]))
            elif self.robot_uids == "xmate3_robotiq":
                qpos = np.array([0, 0.6, 0, 1.3, 0, 1.3, -1.57, 0, 0])
                qpos[:-2] += self._episode_rng.normal(
                    0, self.robot_init_qpos_noise, len(qpos) - 2
                )
                self.agent.reset(qpos)
                self.agent.robot.set_root_pose(sapien.Pose([-0.562, 0, 0]))
            else:
                raise NotImplementedError(self.robot_uids)
```

I’ve noticed that in the `_initialize_episode()` function, the initial joint positions (`qpos`) of the Panda arm are randomized by adding noise (`robot_init_qpos_noise`). This noise affects the robot’s joint configuration at the start of each episode, which in turn changes the hand camera pose and therefore causes variation in the extrinsic camera parameters.
Thank you for clarifying—this explains why the extrinsics are not consistent across episodes. 
However, I’ve noticed that the `cam2world_gl` parameter returned by the environment differs from the `cam2world` transform that I calculate directly from the camera extrinsics. Could you explain why this discrepancy occurs? Are there additional transformations applied internally or differences in coordinate conventions between the environment and the extrinsic computation?

**Comment by StoneT2000** [comment]
You can lookup camworld_gl in this codebase to see what transforms we apply.

The _gl means it is following the OpenGL convention.

**Comment by ZQuang2202** [comment]
> You can lookup camworld_gl in this codebase to see what transforms we apply.
> 
> The _gl means it is following the OpenGL convention.

Thank you! As the problem is resolved, I close this issue.
